### DISCUSSION TITLE
RP2040 I2S full duplex passthrough

### DISCUSSION THREAD

**Discussion by OscarTofa** [earlephilhower/arduino-pico#603] [Q&A]
Hello

i have been experimenting for a few days i2s dell'rp2040. 

I would like to know if it is possible to give an example to use i2s ports in this way?

- an i2s in / out port ( full duplex) 
- a small buffer of 128 or 256 samples
- 48KHz
- 16Bit

This would allow you to be able to do some processing in the rp2040 core and send back the processing signal . It would also be nice to be able to eventually dedicate 1 core to audio processing and the other core for other peripherals.

Thank you

**Comment by earlephilhower** [comment]
I was hoping to get a sample like that, so if you get something please consider a PR.

I'd set up an `I2S(INPUT)` and an `I2S(OUTPUT)`.  Probably 3 buffers each with 96 samples per.  Use the `onReceive()` callback to do the processing of all 96 samples and send to the I2S-OUT.

If you do the `I2S.begin()` from `setup1()` or `loop1()` then it'll run all the audio code on the 2nd core.

**Comment by OscarTofa** [comment]
just to better understand how i2s implemented with the pio works
If I want to make an i2s input & output (same sample rate and same number of bits) I have to initialize 2 separate i2s (so a total of 6 pins) or I can use a single i2s peripheral (so 4 pins) ? because inside the file pio_i2s.pio i see two separate functions for i2s_in and i2s_out but i would like to understand if since they share the same clock and number of bits they can share the same pins

**Reply by earlephilhower** [reply]
They need separate sets of pins.  The state machines are not synchronized amongst each other and in general they do not need to be the same sample size or frequency.

**Reply by OscarTofa** [reply]
Hi 

In trying to get the rp2040 to work in full duplex I found this example very well done.

in addition to explaining the operation in detail, there are both the master mode and the slave mode, both bidirectional

Just to let you know https://github.com/malacalypse/rp2040_i2s_example
